### PR TITLE
maintainers: document expectations

### LINKED ISSUE #171557: CONTRIBUTING.md: Add section on becoming a maintainer
I got some feedback after a review on a PR (https://github.com/NixOS/nixpkgs/pull/161404#issuecomment-1117081506) that it's not very clear that adding yourself as a maintainer in maintainers.nix should be a separate commit.

It might be nice to add a small section to CONTRIBUTING explaining the (simple) process, or maybe linking from there to the manual.

### PR DESCRIPTION
**Pull request by zimbatm**
Motivated by https://discourse.nixos.org/t/where-did-you-get-stuck-in-the-nix-ecosystem-tell-me-your-story

Address the uncertainty around maintainers by defining what it
means, what are the expectations and power you get.

Note to reviewers: some of the numbers are arbitrary and I expect would be tuned by experience.

Are there areas or cases that you think should be covered that I missed?

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Review by matthiasbeyer — APPROVED** [review]

**Review by andir — COMMENTED** [review]
I like the proposal. I don't like the fixed time frames as there might be packages that don't need changes every three months or so.. I'd love to see the timeframes phrased a bit more relaxed.

**Comment by 7c6f434c** [comment]
Maybe measure the time in terms of lack of reaction to specific reports of regressions or significant upstream updates?

**Comment by Janik-Haag** [comment]
We could write a script checking if a package is outdated for $timeframe based on repology data and if it doesn't get updated in $timeframe mark it as unmaintained. I already thought about that a bit.

**Comment by andir** [comment]
> We could write a script checking if a package is outdated for $timeframe based on repology data and if it doesn't get updated in $timeframe mark it as unmaintained. I already thought about that a bit.

I'd not go that far. For this PR we should only try to formalize a formal understanding of the role & the responsibilities and not discuss technical solutions for (arguably) social issues.

**Comment by Janik-Haag** [comment]
I'd not go that far. For this PR we should only try to formalize a formal understanding of the role & the responsibilities and not discuss technical solutions for (arguably) social issues.

I'm fine with that but then we should probably make the timeframe less specific for example writing something like package updates should be performed in a timely manner. I think the issue in this case is not just social since there are other reasons for people to not update stuff like f.e. death so some kind of automation that sends a ping after $months seems like a sensible thing.

**Comment by zimbatm** [comment]
Relaxed the wording a little bit so the 3 months are in relation to package-related notifications. If the package is stable it won't generate any notifications (upgrades / issues / ...) and the maintainer would stay on board.

File: maintainers/README.md
Comment (RaitoBezarius) [inline maintainers/README.md:14]: I would add a mention for packages that are not affected by the "one week" rule, that is:  "code owned" packages which have a large reverse dependency tree, e.g. systemd, LLVM, etc.

I don't think it's easy to put an amount of time, I'd state it as:

> Committers should reach out to the active code owners and discuss the timing for a potential merge, in the event where no active code owner can be reached in a reasonable amount of time (i.e. taking into account potential holidays), merge can be performed nevertheless with a provided rationale.

The idea is to:

(1) ensure that no stray committer merges a systemd change while not being in touch with active code owner of e.g. systemd which may have extra context on ongoing changes (internal information, etc.)

(2) ensure that abandoned "important dependencies" can be slowly picked up by new committers who will garden them and find a new maintenance story for them hopefully, e.g. LLVM

Tiny version bumps and security updates _could_ not be totally exception here, I don't have a strong opinion on that, I know that systemd security updates can lead to breakage, we had e2fs tooling's tiny version bumps breaking bootloaders (something something XFS), etc. It's a very hard problem in general.

**Review by RaitoBezarius — APPROVED** [review]
This is very cool, thank you so much!

**Review by JulienMalka — APPROVED** [review]
I like this proposal, thank you!

**Review by Shawn8901 — APPROVED** [review]
I like it. 👍

**Review by infinisil — APPROVED** [review]
I wish we had better automation for this kind of thing, but documenting this (and adding some rough timings) is already an improvement!

Let's include the suggestion from https://github.com/NixOS/nixpkgs/pull/250344#discussion_r1299443711 though, I'll merge it after that.

File: maintainers/README.md
Comment (Ma27) [inline maintainers/README.md:39]: I assume that "no reaction at all" is meant here, correct? I.e. this isn't done on a per-package basis (otherwise one could lose the status if a fellow maintainer is always faster and you don't have anything to add most of the time).

When I first read this paragraph, I wasn't sure, so perhaps this should be rephrased a bit to make this clearer.

Also, we'd to give people who are about to lose their status a heads-up, correct? I  guess this should be mentioned here as well :)

File: maintainers/README.md
Comment (Ma27) [reply maintainers/README.md:14]: > Tiny version bumps and security updates could not be totally exception here, I don't have a strong opinion on that, I know that systemd security updates can lead to breakage, we had e2fs tooling's tiny version bumps breaking bootloaders (something something XFS), etc. It's a very hard problem in general.

Would it make sense to make an exception for security updates / fixes earlier if e.g. the security team is involved as reviewer?

Generally I think that case of e.g. important bugfixes (e.g. $bump fixes data loss on runtime) or security fixes it should be possible for a committer to decide on a case-by-case basis what's best. @RaitoBezarius has brought some great examples on why I doubt that it's possible to find a general rule for this.

Also, such a ruling would exclude all those cases where the sole motivation of a person not maintaining the package in question is e.g. "wants to have the latest version available, immediately".

Comment (fricklerhandwerk) [reply maintainers/README.md:14]: Agreed with @RaitoBezarius on specifying timing, or rather pushing that responsibility to maintainers themselves. Maybe we could establish a convention where maintainers are supposed to communicate what they can deal with.

But in general, why would random people merge code owned by others at all?

File: maintainers/README.md
Comment (zimbatm) [reply maintainers/README.md:39]: Removing a maintainer requires making a PR so that that person would be notified. But I'll add a paragraph for that.

I still think the test should be per package. Some people are active in some areas of nixpkgs, and have abandoned other parts that they don't care about anymore.

Worst case you can react on the removal PR.

Comment (zimbatm) [reply maintainers/README.md:38]: 10e3f2c7df0c50f47bdbb3029d475642c77b1246

File: maintainers/README.md
Comment (zimbatm) [reply maintainers/README.md:14]: Can you expand on what you mean? I'm not sure to understand

Comment (7c6f434c) [reply maintainers/README.md:14]: «By default, we expect committers to wait at least a week before merging changes on packages they are not maintaining.» — however, we do not currently wait to give _all_ maintainers time to comment, if one non-committer maintainer asks to merge, this is currently supposed to be good enough when the committer also likes the change.

File: maintainers/README.md
Comment (zimbatm) [reply maintainers/README.md]: Can you look at c67b737cd4d81303dc9cdaeb6f7b6f4be42a2ae7 and tell me if that is good enough? I'm keeping the text short so people are more likely to read it. And therefor try to not overload with exceptions.

**Comment by zimbatm** [comment]
I believe this PR to be in good shape, with the eventual typos. Please take another look.

File: maintainers/README.md
Comment (7c6f434c) [inline maintainers/README.md:21]:
```suggestion
changes not endorsed by a package maintainer (which may be themselves). This should leave enough time
```

File: maintainers/README.md
Comment (zimbatm) [reply maintainers/README.md]: I also realize we don't describe how conflict resolution happens in case multiple maintainers of the same package disagree.

File: maintainers/README.md
Comment (ajs124) [inline maintainers/README.md:21]: Does that mean the endorsement of one maintainer is enough? What if a package has multiple maintainers? What if they disagree?

Comment (zimbatm) [reply maintainers/README.md:21]: I don't know if I'm ready to open that rabbit hole. For now, I suggest leaving that unspecified.

Comment (RaitoBezarius) [reply maintainers/README.md:21]: IMHO, conflict resolution is 2 or 3 magnitudes harder than anything those PRs are doing so far, I would +1 not going into that rabbit hole.

File: maintainers/README.md
Comment (RaitoBezarius) [reply maintainers/README.md]: Interesting point @ architectures they are willing to maintain.

File: maintainers/README.md
Comment (7c6f434c) [reply maintainers/README.md:21]: I think putting a specific duration on «wait for some maintainer reaction» is good, and everything further is reliant on judgement calls that are hard to predefine. We don't ask to definitely merge once there is a single maintainer approval, after all!

File: maintainers/README.md
Comment (RaitoBezarius) [reply maintainers/README.md]: I opened https://github.com/NixOS/nixpkgs/pull/251008 for exploring the question, feel free to carry the conversation there.

File: maintainers/README.md
Comment (fricklerhandwerk) [inline maintainers/README.md:28]:
```suggestion
In case of critical security updates, the [security team](https://nixos.org/community/teams/security) might override these
```

File: maintainers/README.md
Comment (fricklerhandwerk) [inline maintainers/README.md:26]:
```suggestion
author is listed in the [`CODEOWNERS`](../.github/CODEOWNERS) file.
```

File: maintainers/README.md
Comment (fricklerhandwerk) [inline maintainers/README.md:37]:
```suggestion
maintainer. Commit access to the Nixpkgs repository is not required for that.
```

File: maintainers/README.md
Comment (fricklerhandwerk) [inline maintainers/README.md:45]:
```suggestion
Maintainers who have become inactive on a given package can be removed. This
```

File: maintainers/README.md
Comment (fricklerhandwerk) [inline maintainers/README.md:47]:
```suggestion
helps us keep an accurate view of the state of maintenance in Nixpkgs.
```

File: maintainers/README.md
Comment (ajs124) [reply maintainers/README.md:21]: Alright, then lets postpone this. It's just something I'm regularly asking myself when trying to decide if I should just merge something or if maybe I should wait.

Comment (7c6f434c) [reply maintainers/README.md:21]: I guess for me the correct question will now be «will I be sincere or hypocritical saying ‘Sorry, after a single maintainer approval I did not understand this is going to be controversial’?».

File: maintainers/README.md
Comment (AndersonTorres) [inline maintainers/README.md:41]: While reviewing #250527, I noticed the advice for creating a separate commit for `maintainers` is neither documented in Contributing nor Maintainers files

https://github.com/NixOS/nixpkgs/pull/250527#issuecomment-1693276228

Should we include it here?

**Review by AndersonTorres — CHANGES_REQUESTED** [review]
Also, I suggest something about reading the manuals.

**Comment by zimbatm** [comment]
I'm sorry, but this last review is expanding the scope of the PR. This is exactly how PRs die; because reviewers keep asking for more stuff until the contributor gets exhausted.

Before submitting reviews please ask yourself; Is this a unit of work that is consistent and that improves on the existing situation? If yes then let's merge this and start gathering real-world data from it. I'm sure more PRs will follow afterwards.

File: maintainers/README.md
Comment (Janik-Haag) [reply maintainers/README.md:41]: There was another PR by @matthiasbeyer adding that IIRC

**Review by AndersonTorres — APPROVED** [review]

File: maintainers/README.md
Comment (SuperSandro2000) [reply maintainers/README.md:21]: Can we have some tooling to get reminded of that time?

File: maintainers/README.md
Comment (SuperSandro2000) [inline maintainers/README.md:60]: * that currently does not work all the time and has edge cases where it breaks which is unfortunate

File: maintainers/README.md
Comment (SuperSandro2000) [inline maintainers/README.md:22]: That is not true for all subsystems. For example the python subsystem is regularly ignoring this on purpose to keep a healthy ecosystem. I am not sure how this alignes with this now.

File: maintainers/README.md
Comment (RaitoBezarius) [reply maintainers/README.md:21]: I do not want to be mean, but calendars are pretty good for this.

File: maintainers/README.md
Comment (RaitoBezarius) [reply maintainers/README.md:60]: It's not a big deal, this is a OfBorg bug.

File: maintainers/README.md
Comment (RaitoBezarius) [reply maintainers/README.md:22]: Can you please be clearer about what you mean by this? e.g. Python committers merging bump < 1week because we want to keep up-to-date packages? I do not see why a 1-week wait could not be honored. 

It would just create a 1-week lag, which is not a big deal IMHO.

Comment (7c6f434c) [reply maintainers/README.md:22]: > ignoring this on purpose to keep a healthy ecosystem

Can this be described as «python team considers selves as comaintainers of all the packages inside `python-packages.nix`»?

> It would just create a 1-week lag, which is not a big deal IMHO.

Adding a 1-week lag to a process that works without it needs a good justification why the process will start working better afterwards…

File: maintainers/README.md
Comment (7c6f434c) [reply maintainers/README.md:21]: (Actually, no, they are not, for _multiple_ reasons: they are hard to set up to get exactly the relevant notifications for this batch when you cut out some time to look at Nixpkgs, and most of them don't integrate well enough with browsing the pull request. On the other hand, an option for CI re-requesting reviews in case there is a check you want to let finish is useful in any case, and if it exists, maybe remind-me-in-a-week would be doable on top of that)

File: maintainers/README.md
Comment (RaitoBezarius) [reply maintainers/README.md:22]: > Adding a 1-week lag to a process that works without it needs a good justification why the process will start working better afterwards…

Well, it all boils down again to the position of the package in the dependency tree, some folks do not take this into account and just go on a happy merge spree, some cares about, etc.

Either case, @mweinelt has probably good opinions on that and I trust their judgement on those questions.

Comment (RaitoBezarius) [reply maintainers/README.md:22]: BTW,

> Can this be described as «python team considers selves as comaintainers of all the packages inside python-packages.nix»?

Yes, it would be good to ask the opinion of the Python team before writing something they may not agree about. :)

cc @FRidh @jonringer @tjni

Comment (mweinelt) [reply maintainers/README.md:22]: The python ecosystem only ever carries one* specific version for each package, so we often have to touch many packages, to keep everything working. I would guess that only 10% of all 6000+ python packages have active maintainers, so bulk updates have become a requirement for a healthy package set. Ultimately this situation led to a few very active maintainers essentially co-maintaining most of the package set.


<sup>* yes, there are exceptions, but they're exactly that<sup>

Comment (tjni) [reply maintainers/README.md:22]: We definitely don't optimize for giving maintainers a chance to respond, because the definition of maintainer for most Python packages has been: "I added the package, I might be interested in keeping it up to date, and I might be able to help answer some questions about it."

Comment (RaitoBezarius) [reply maintainers/README.md:22]: Either case, the mentioned sentence is "for most packages", I would argue: we should get this in and refine it for Python ecosystems and similar ecosystems.

File: maintainers/README.md
Comment (zimbatm) [reply maintainers/README.md:21]: I agree; we need some more tooling for this.

File: maintainers/README.md
Comment (FRidh) [inline maintainers/README.md:24]:
```suggestion
For critical packages, this convention needs to be negotiated with the
```

**Review by FRidh — COMMENTED** [review]
Looks good to me.

File: maintainers/README.md
Comment (zimbatm) [reply maintainers/README.md:22]: Let's keep this in mind for after the PR is merged. We need a better mechanism for maintainers to communicate those expectations.

**Comment by zimbatm** [comment]
Alright, this is ready. Unless anything new comes up, I'll merge it in a few days.

**Review by infinisil — APPROVED** [review]

**Pull request merged**
